Metrics: Add version metric

ChristineTChen opened the pull request:
Adds `cilium_version` prometheus metric that is labelled by the Cilium version.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](http://docs.cilium.io/en/stable/contributing/contributing/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue. 
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](http://docs.cilium.io/en/stable/contributing/contributing/#dev-coo).
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
Add cilium_version metric
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9623)
<!-- Reviewable:end -->

joestringer commented on pkg/version/version.go:
```diff
@@ -57,7 +59,9 @@ func FromString(versionString string) CiliumVersion {
 func GetCiliumVersion() CiliumVersion {
-	return FromString(Version)
+	cver := FromString(Version)
+	metrics.VersionMetric.WithLabelValues(cver.Version)
```
As far as I can tell, most other usage of `WithLabelValues()` is followed by a verb function like `Observe()`, `Add()`, `Inc()` but this one doesn't. My reading of the `WithLabelValues()` function comment is that this will result in creating the label with the value 0:

> // It is possible to call this method without using the returned Gauge to only  
> // create the new Gauge but leave it at its starting value 0. See also the      
> // SummaryVec example.

Does this mean that the gauge shows up with a label named with the version, or with a label named `version` with empty contents, or does it actually show up with `version` = `vX.Y.Z`?

Also, I'm wondering if `GetCiliumVersion()` is the best place to put the metric increment or if it should be configured somewhere else (from the main daemon/ files somewhere?). It seems like this function is only called from the CLI, rather than from the main daemon binary:

```
$ git grep "GetCiliumVersion()"
cilium/cmd/version.go:  return version.GetCiliumVersion()
pkg/version/version.go:func GetCiliumVersion() CiliumVersion {
```

raybejjani replied:
Agreeing with joestringer, perhaps it would be better to set this in the metrics init functions. Either `init` itself (https://github.com/cilium/cilium/blob/master/pkg/metrics/metrics.go#L1054) or when we create the configuration (https://github.com/cilium/cilium/blob/master/pkg/metrics/metrics.go#L503).
More generally, we do have a `process_start_time_seconds` metric but it is a prometheus built-in (https://github.com/cilium/cilium/blob/master/pkg/metrics/metrics.go#L1055). If we could figure out how to label that it would be a natural spot for this info (on the assumption that the goal is to know when cilium-agents have upgraded etc. `process_start_time_seconds` would also give you when that instance started).

ChristineTChen replied:
Hi @joestringer! Apologies for the confusion, I misunderstood the docs for `WithLabelValues`. My intention was to submit a dummy metric that is labelled `cilium_version=x.y.z`.

If the metric should be placed elsewhere, I'd be more than happy to move it to somewhere more appropriate. My intention is to submit as a prometheus metric. I'm not sure how helpful it would be to display the metric in the CLI as `cilium version` is already an option.

Thank you! 😄

ChristineTChen replied:
@raybejjani, thanks for the feedback! I'll update accordingly and move the metric.

joestringer reviewed (commented):
Hi, thanks for the contribution!

I'm trying to figure out how to test this, is this expected to show up in `cilium metrics list` as well or just prometheus metrics?

raybejjani reviewed (changes requested):
Thanks for the PR :)
If you get a chance, could you squash the second commit (and any others that are just fixups) before the PR is merged?

raybejjani reviewed (approved):
Works for me!

raybejjani reviewed (changes requested):
@ChristineTChen I think the `VersionMetric.WithLabelValues(version.GetCiliumVersion().Version)` is in the wrong spot. It happens before the metrics instance is created, and so used the no-op dummy. Could you move it to where we create the version metric?

ChristineTChen committed:
Add Prometheus version metric

The version metric is labelled by the Cilium version.

Signed-off by: Christine Chen <christine.chen@datadoghq.com>

joestringer reviewed (approved):
:+1: I see it show up in `cilium metrics list` now:

```
$ cilium metrics list | grep cilium_version
cilium_version                                        version="1.6.90"                                                                                      0.000000
```
